### ISSUE TITLE
geoip filter does not create elasticsearch mapping with type "geo_point"

### ISSUE THREAD

**Issue by Poohblah** [elastic/logstash#3137]
I'm attempting to use the [geoip filter](http://logstash.net/docs/1.4.2/filters/geoip) to populate my logs with geographic coordinate data that I can use to drive the "tile map" visualization in Kibana 4.  The problem is that the data retrieved by this filter is mapped as a "double" type instead of a "geo_point" type, which prevents the tile map from recognizing the data as a geographic coordinate data source.

According to the docs:

> Starting with version 1.3.0 of Logstash, a [geoip][location] field is created if the GeoIP lookup returns a latitude and longitude. The field is stored in GeoJSON format. Additionally, the default Elasticsearch template provided with the elasticsearch output maps the [geoip][location] field to a geo_point.

However, the field `[geoip][location]` is a double for me, not a geo_point. Here is the mapping, along with my logstash configs, to demonstrate the issue:
```
[welastic ~]% curl 'localhost:9200/access-2015.04.30/_mapping/access?pretty=true'
{
  "access-2015.04.30" : {
    "mappings" : {
      "access" : {
        "properties" : {

[snip]

          "geoip" : {
            "properties" : {
              "area_code" : {
                "type" : "long"
              },
              "city_name" : {
                "type" : "string"
              },
              "continent_code" : {
                "type" : "string"
              },
              "country_code2" : {
                "type" : "string"
              },
              "country_code3" : {
                "type" : "string"
              },
              "country_name" : {
                "type" : "string"
              },
              "dma_code" : {
                "type" : "long"
              },
              "ip" : {
                "type" : "string"
              },
              "latitude" : {
                "type" : "double"
              },
              "location" : {
                "type" : "double"
              },
              "longitude" : {
                "type" : "double"
              },
              "postal_code" : {
                "type" : "string"
              },
              "real_region_name" : {
                "type" : "string"
              },
              "region_name" : {
                "type" : "string"
              },
              "timezone" : {
                "type" : "string"
              }
            }
          },

[snip]

        }
      }
    }
  }
}
```

Here is the relevant logstash filter configs:
```
[wlogstash ~]% cat /etc/logstash/conf.d/2-filter.conf 
filter {

  if [type] == "access" {

[snip]

    geoip {
        source => "client_ip"
    }

[snip]

  }
}
```

And here is the relevant output config:
```
[wlogstash ~]% cat /etc/logstash/conf.d/3-output.conf
output {
  if [type] == 'access' {
    elasticsearch {
      index => "access-%{+YYYY.MM.dd}"
      host => "welastic.********.***"
    }

[snip]

  }
}
```

I'm using logstash 1.4.2 from the logstash rpm repository:
```
[wlogstash ~]% rpm -q logstash
logstash-1.4.2-1_2c0f5a1.noarch
```
against an elasticsearch 1.5.2 back-end, also from the elasticsearch rpm repository:
```
[welastic ~]% rpm -q elasticsearch
elasticsearch-1.5.2-1.noarch
```

**Comment by jordansissel** [comment]
The problem is your index name is "access-...", if you look at the template, it only matches [logstash-*](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/elasticsearch-template.json#L2).

We are exploring ways to have logstash use the `index` setting on the elasticsearch output to inform what the index pattern should be in the template, but this isn't easy. There's a patch from @Bradox (https://github.com/logstash-plugins/logstash-output-elasticsearch/pull/25) that works towards this, if you want to try it.

----

For today, there are two possible workarounds:

1) As a workaround, you'll want to create your own template (which can be based on the [default one](https://github.com/logstash-plugins/logstash-output-elasticsearch/blob/master/lib/logstash/outputs/elasticsearch/elasticsearch-template.json)) that matches your index naming (`access-*`). 

or 2) An alternate workaround would be either use the default index naming (logstash-YYYY.mm.dd) or use `logstash-` as a prefix, such as `logstash-access-%{+YYYY.MM.dd}`.

**Comment by Poohblah** [comment]
Thanks, Jordan!  I renamed my index to "logstash-access" and this has fixed the problem. :)

This behavior wasn't clear to me from the documentation.  The only related bit I saw in the docs was for the `template` field in the elasticsearch output:

> You can set the path to your own template here, if you so desire. If not set, the included template will be used.

This led me to believe that the default template would be always be applied unless I changed the value of this configuration option. I guess this may be *technically* true, but I didn't realize that the default template was dependent on the name of my indices as well.

Digging a bit deeper into the docs, I see this for `manage_template`:

> a default mapping template for Elasticsearch will be applied, if you do not already have one set to match the index pattern defined (default of “logstash-%{+YYYY.MM.dd}”), minus any variables. For example, in this case the template will be applied to all indices starting with logstash-*

So I guess that this behavior is documented.

I'll close this issue now, since there isn't actually any bug, just my own trouble in interpreting the documentation...

Thanks again.

**Comment by jordansissel** [comment]
I totally agree the behavior is confusing: Setting `index` and then Logstash still tries to manage a template that matches something that isn't your `index`! Not ideal, for sure, even if we document it :)

We'll try to fix it so that it'll work better in the future, but I'm glad we were able to clear this up for you. :)

**Comment by gabrielhao** [comment]
May I ask if this issue is fixed in later version? like 1.5.x

**Comment by dfaropennetwork** [comment]
My resolution for help the people

Thax Dmitry the problems was on the index template, i needed one and i had to select manual mapping for geoip fields, but all this problems are resolved in the elasticsearch version 2.0, instead i put my index mapping for help the people with old versions, 1.4.* etc...

Template.json

{
  "order" : 0,
  "template" : "mytemplate-*",
  "settings" : { "index.refresh_interval" : "5s" "number_of_shards": "2"},
  "mappings" : {
    "_default_" : {
      "dynamic_templates" : [
        {
          "message_field" : {
            "mapping" : { "index" : "analyzed", "omit_norms" : true, "type" : "string" },
            "match_mapping_type" : "string",
            "match" : "message" }
          },
        {
          "string_fields" : {
            "mapping" : { "index" : "not_analyzed", "ignore_above" : 256, "type" : "string" },
            "match_mapping_type" : "string",
            "match" : "*" }
          }
       ],
      "properties" : {
        "geoip" : {
          "dynamic" : true,
          "path" : "full",
          "properties" : { "location" : { "type" : "geo_point" } },
          "type" : "object" },
        "@version" : { "index" : "not_analyzed", "type" : "string" }
      },
      "_all" : { "enabled" : true }
    }
  },
  "aliases" : { }
}


And the logstash configuration very important, file 30-lumberjack-output.conf for example

output {
  elasticsearch {
  host => localhost
  index => "mytemplate-%{+YYYY.MM.dd}"
  template => "/etc/logstash/mytemplate.json"
  template_name => "mytemplate-*"
  manage_template => "true"
  template_overwrite => "true"
}
  stdout { codec => rubydebug }
}

**Comment by dfaropennetwork** [comment]
The problem are fixed in the elasticsearch 2.0 version

**Comment by chrisribe** [comment]
Thanks dfaropennetwork your solution with the template name was my problem for geo_point.
The documentation is not clear / does not have simple examples of a custom template.

I found this apache example that I was able to get working at first.
https://github.com/elastic/examples/tree/master/ElasticStack_apache

But it only shows it working for a bulk import, in reality access logs get updated hence the (in my case) apache_access-* index. It never occured to me that the template name NEEDS to match the index name !!
